Resources/doc/2-configuration.md
## Configuration

This bundle can be configured, and this is the list of what you can do :
* Create many pheanstalk client.
* Define specific server / host for each connection.
* Define specific port for each connection. This option is optional and default value is 11300.
* Define specific timeout for each connection. Timeout refere to the connection timeout. This option is optional and default value is 60.
* Use custom proxy for pheanstalk client.
* Disable this bundle. This options is optional and default value is true. 

``` yaml
# app/config/config.yml
leezy_pheanstalk:
    enabled: true
    connection:
        primary:
            server: beanstalkd.domain.tld
            port: 11300
            timeout: 60
        secondary:
            server: beanstalkd-2.domain.tld
            default: true
            proxy: acme.pheanstalk
```

*acme.pheanstalk* is a custom proxy which implements the *Leezy\PheanstalkBundle\Proxy\PheanstalkProxyInterface* interface.

**Note:**
```
    You can retreive each pheanstalk using the container with "leezy.pheanstalk.[pheanstalk_name]".
    When you define a "default" pheanstalk. You can have a direct access to it with "leezy.pheanstalk".
```

``` php
<?php

namespace Acme\DemoBundle\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\Controller;

class HomeController extends Controller {

    public function indexAction() {
        $pheanstalkPrimary = $this->get("leezy.pheanstalk.primary");
        $pheanstalkSecondary = $this->get("leezy.pheanstalk");

        // ----------------------------------------
        // producer (queues jobs) on beanstalk.domain.tld

        $pheanstalkDefault
            ->useTube('testtube')
            ->put("job payload goes here\n");

        // ----------------------------------------
        // worker (performs jobs) on beanstalk-2.domain.tld

        $job = $pheanstalkSecondary
            ->watch('testtube')
            ->ignore('default')
            ->reserve();

        echo $job->getData();

        $pheanstalkSecondary->delete($job);

        // ----------------------------------------
        // on each defined pheanstalks
        $pheanstalkLocator = $this->get("leezy.pheanstalk.pheanstalk_locator");

        foreach($pheanstalkLocator->getPheanstalks() as $pheanstalk) {
            $pheanstalk
                ->useTube('boardcast')
                ->put("job payload goes here\n");
        }
    }

}
?>
```


CHANGELOG
== 2.1.2 ==
* Add KickJobCommand
* Fix error management on command

== 2.1.1 ==
* Fix missing implementation of PheanstalkInterface::kickJob

== 2.1.0 ==
* Use pda/pheanstalk 2.1.0
* Add KickCommand
* Fix default name on Command

== 2.0.1 ==
* Set type hint for logger
* Add few tests on PheanstalkLogListener

== 2.0.0 ==
* Add PheanstalkProxy
* Allow user to define custom PheanstalkProxy
* Fragment documentation
* Events documentation
* Change 'connection' name to 'pheanstalk'
* Event integration
* BC : Use 'pheanstalk' name instead of 'connection' to avoid confusion
* Add more security on CLI
* Logger integration
* Add documentation on custom proxy

== 1.2.0 ==
* Add profilter integration, thanks @Maxwell2022
* Add ConnectionLocator

== 1.1.0 ==
* Remove useless Tests/Fixtures/Entity/FakeMedia.php
* Update composer to use symfony/framework-bundle instead of symfony/symfony
* Remove requirement PHP >= 5.3.3. Use the PHP requirement of symfony/framework-bundle
* Fix Travis-CI problem

== 1.0.4 ==
* Update to PSR-0 : https://github.com/pda/pheanstalk/commit/412ab5b7469538ab613efd8fb91b4adc84bf599f
* Add warning about composer minimum-stability due to : https://github.com/armetiz/LeezyPheanstalkBundle/issues/8

== 1.0.3 ==
* Update code to pda/pheanstalk which is a PHP5.2 implementation.
* BC still ok.

== 1.0.2 ==
* Add Unit Test
* Throw exception when more than one connection have the default flag
* Switch dependency from 'mrpoundsign/pheanstalk-5.3' to 'pda/pheanstalk'
* Add TravisCI configuration

== 1.0.1 ==
* Fix dependency to pheanstalk-5.3. #4

== 1.0.0 ==
* Update README to be composer compliant.
* First stable tag version.

DependencyInjection/Configuration.php
<?php

namespace Leezy\PheanstalkBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * This is the class that validates and merges configuration from your app/config files
 *
 * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html#cookbook-bundles-extension-config-class}
 */
class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritDoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode = $treeBuilder->root('leezy_pheanstalk')->children();

        $rootNode
            ->arrayNode('profiler')
                ->addDefaultsIfNotSet()
                ->children()
                    ->scalarNode('enabled')->defaultValue('%kernel.debug%')->end()
                    ->scalarNode('template')->defaultValue('LeezyPheanstalkBundle:Profiler:pheanstalk.html.twig')->end()
                ->end()
            ->end()
            ->booleanNode("enabled")->defaultTrue()->end()
            ->arrayNode("pheanstalks")
                ->requiresAtLeastOneElement()
                ->useAttributeAsKey('name')
                ->prototype('array')
                    ->children()
                        ->scalarNode("server")
                            ->isRequired()
                            ->cannotBeEmpty()
                        ->end()
                        ->scalarNode("port")
                            ->cannotBeEmpty()
                            ->defaultValue("11300")
                        ->end()
                        ->scalarNode("timeout")
                            ->cannotBeEmpty()
                            ->defaultValue("60")
                        ->end()
                        ->booleanNode("default")
                            ->defaultFalse()
                        ->end()
                        ->scalarNode("proxy")
                            ->cannotBeEmpty()
                            ->defaultValue("leezy.pheanstalk.proxy.default")
                        ->end()
                    ->end()
                ->end()
            ->end()
        ->end();

        return $treeBuilder;
    }
}
